### PR TITLE
Adds tasks/TODOs support

### LINKED ISSUE #308: Task (TODO) support
Like other people, I like this widget, but I really want a widget that displays calendar events and tasks (or TODOs) together. Like Pure Calendar Widget does (or did), but unfortunately it's not working reliably lately and it seems not to be maintained anymore.

I read the previous issues about this topic, including the differently colored past events, but I find that solution unappealing.

Unfortunately, Android does have a concept of tasks as it does of calendar events. So it's not possible to have an universal solution that would just work with any of the dozens of Task applications available on the Play Store. Instead, code needs to be added for each application that would be supported by the widget.

I'm opening this new issue because I'm willing to write this code, if you'd be interested in including it in your app. Initially I'd add support for Samsung Calendar (because that's what I use and I've already reverse engineered how to fetch its tasks), but I'd try to write the code in a way that it should be simple to add support for other application. Another that comes to mind is  https://github.com/dmfs/opentasks, it's open source and it's interfaces are available.

Anyway, let me know if you think this would be a nice addition to this widget.

### PR DESCRIPTION
**Pull request by ekalin**
Fixes #308.

This is still somewhat basic, but functional, fetching tasks from Samsung Calendar or Dmfs OpenTasks, and it should be relatively straightforward to add support for other applications (as long as one knows how to fetch data from its content provider and set up an Intent to open the task in that application).

### CODE REVIEW

File: app/src/androidTest/java/org/andstatus/todoagenda/BirthdayTest.java [hunk 1]
```diff
@@ -19,31 +18,34 @@
 
     private static final String TAG = BirthdayTest.class.getSimpleName();
 
-    private MockCalendarContentProvider provider = null;
+    private MockWidget widget = null;
+    private MockContentProviderImpl provider = null;
```
Comment (yvolk) [inline app/src/androidTest/java/org/andstatus/todoagenda/BirthdayTest.java]: I think splitting MockCalendarContentProvider into two classes is not necessary, it only increases number of lines to use it and causes lots of unneeded changes in tests thus making it much harder to see the essence of the proposed change :-(
Moreover, "MockWidget" and widget seem to me less fitting to their real meaning...

File: app/src/main/java/org/andstatus/todoagenda/WidgetInfoStorage.java [hunk 2]
```diff
@@ -25,16 +22,12 @@
 /**
  * @author yvolk@yurivolkov.com
  */
-public class CalendarQueryResultsStorage {
+public class WidgetInfoStorage {
```
Comment (yvolk) [inline app/src/main/java/org/andstatus/todoagenda/WidgetInfoStorage.java]: **WidgetInfo**Storage name is much less concrete  than Calendar**QueryResults**Storage thus making purpose of the class harder to understand. This is all about "widget" and "info" here...

**Comment by ekalin** [comment]
You're right, there are some big changes. I was aware of that when I made them, but I disagree that 90% of what's changed is just restructuring, and that it's unnecessary. I do think that they are important, and adding the task support without them might be possible, but would leave the code hard to maintain.

I do agree that the names suck. :-( I couldn't come up with anything better, but I wasn't really satisfied with them.

I'll add some more comments in the review comments themselves.

File: app/src/main/java/org/andstatus/todoagenda/WidgetInfoStorage.java (hunk 2 above)
Comment (ekalin) [reply app/src/main/java/org/andstatus/todoagenda/WidgetInfoStorage.java]: That's true, but even before, it already stored more than the query results - it stores settings and some device information used for debugging. And now it stores also task results, so while WidgetInfo may be a bad name, I think CalendarQueryResults is also somewhat misleading.

File: app/src/androidTest/java/org/andstatus/todoagenda/BirthdayTest.java (hunk 1 above)
Comment (ekalin) [reply app/src/androidTest/java/org/andstatus/todoagenda/BirthdayTest.java]: Here I have to disagree with you. MockCalendarContentProvider did much more than being a mock content provider for calendar - it was a whole test harness that creates a Context, simulates some necessary features of the widget (loading state, editing settings, fetching and building the event list, etc).

And on a practical note, I needed a new mocked ContentProvider for tasks - there's even a test that uses calendar and tasks at the same time, and splitting only the ContentProvider part allowed that class to be reused, and the test harness part to be used in the new task tests.

For what's worth, the changes to existing tests were basically just the refactorings - as far as I remember it was not necessary to make any significant changes to them.

Comment (yvolk) [reply app/src/androidTest/java/org/andstatus/todoagenda/BirthdayTest.java]: The problem is that I don't like these refactoring :-(
As current app maintainer I assure you that having your PR with addition of the new feature without any refactoring will help future app maintainance greatly.
Just extend existing code, I will refactor everything later. Please.
And please squash all your changes into one commit.

File: app/src/main/java/org/andstatus/todoagenda/WidgetInfoStorage.java (hunk 2 above)
Comment (yvolk) [reply app/src/main/java/org/andstatus/todoagenda/WidgetInfoStorage.java]: No, the class doesn't store settings, there is org.andstatus.todoagenda.prefs.SettingsStorage for that.
It doesn't store device information either: it only adds that information, when we "shareEventsForDebugging"

**Comment by ekalin** [comment]
I'll make the changes, but I'm somewhat busy lately. It'll take a few days, but I have not forgotten this PR. :-)

**Comment by yvolk** [comment]
@ekalin Thank you. 
And then we will talk about Tasks :-)

**Comment by ekalin** [comment]
OK, I've undone the renaming of CalendarQueryResultsStorage, and removed the tests refactoring.

As a side effect, the new tests are also gone :-) But I can work on some tests if this is going to be accepted. As a matter of fact, most of the tests could be done as simple unit tests (the exeception would be one test including calendar events and tasks, for this an integration test is valid), but due to the lack of other unit tests I suppose this would not be a viable option, right?

I haven't squashed the commits yet. If that's OK I'd rather squash them as a last step before merging, but if you really want I can do that now.

**Comment by yvolk** [comment]
@ekalin In order to help us move further I continued cleaning of your change, fixed a couple of bugs and pushed the whole change as the single commit:
https://github.com/andstatus/todoagenda/commit/64bac60dc054d8270f39b5be29c2d0e4300954fd
Two things to fix are there yet:
1. Add Tasks permission(s) to PermissionsUtil (I removed your change in the settings activity for unity)
2. Fix probable bug: org.andstatus.todoagenda.prefs.InstanceSettings#getTaskSource is unused ???

**Comment by ekalin** [comment]
I see that you're storing calendar and tasks together in CalendarQueryResultsStorage. This might make it complicated in MockCalendarContentProvider if it were necessary to create a test that includes calendar and tasks, it would need to look at the uri to differentiate each kind.

That little hack in WidgetConfigurationActivity is necessary so that the "Request permission" button disappears from the settings page when the permission is granted. It's a little inelegant, but I think any good solution would require changing to the Support Library's PreferenceActivity.

Which brings me to your TODO #1: I'm not sure what you expect. It's likely that each Task application will require a different permission to read its tasks, so the permission cannot be in PermissionsUtil.

InstanceSettings#getTaskSource ended unused; I'll remove it.

**Comment by yvolk** [comment]
> I see that you're storing calendar and tasks together in CalendarQueryResultsStorage. This might make it complicated in MockCalendarContentProvider if it were necessary to create a test that includes calendar and tasks, it would need to look at the uri to differentiate each kind.

That's right. We will need to differentiate each result by Event Provider, and we have them more than two already (Android Calendar, Open Tasks, Samsung tasks...)

**But let's do this at the next step**. Here we are only cleaning the code from unneeded changes and getting rid of easy to see bugs.

> That little hack in WidgetConfigurationActivity is necessary so that the "Request permission" button disappears from the settings page when the permission is granted. It's a little inelegant, but I think any good solution would require changing to the Support Library's PreferenceActivity.

Exactly the same problems with Calendar permission are solved by usage of PermissionsUtil methods. "Little hack" for each button/permission is not only dirty: it doesn't cover all problems not having the permission. In different parts of the app, not in the setting only.

> Which brings me to your TODO #1: I'm not sure what you expect. It's likely that each Task application will require a different permission to read its tasks, so the permission cannot be in PermissionsUtil.

Different permissions for different widgets (if you have more than one), This is why PermissionsUtil takes (should take) settings of concrete widget as an argument.

> InstanceSettings#getTaskSource ended unused; I'll remove it.

It **should** be used, because TaskSource may be different for different widgets on the same device! Create two widgets with different task sources and see what you get! (This is why tests are needed...)

**Comment by yvolk** [comment]
@ekalin Please continue your changes with my commit, so we wouldn't need to always compare different branches and rebase on master...

**Comment by ekalin** [comment]
Funny you agree about separating calendar events and tasks in QueryResultStorage, because that's what my code did. But never mind.

I've fixed the problem with two widgets with different TaksSources.

As for the permissions issue, I'm afraid I don't understand what's your idea about what should be done.

**Comment by yvolk** [comment]
> Funny you agree about separating calendar events and tasks in QueryResultStorage, because that's what my code did. But never mind.

Yes, we need the separation, but not via creation of a separate List with tasks' results... I will do this.

> I've fixed the problem with two widgets with different TasksSources.

Thank you.

> As for the permissions issue, I'm afraid I don't understand what's your idea about what should be done.
I will fix this also and show you.

I think that at this stage we can merge this PR. I will do additional fixes as separate commits and let you know.

**Pull request merged**